### DISCUSSION TITLE
How to move the router state to store?

### DISCUSSION THREAD

**Discussion by sandrig** [pmndrs/zustand#2241] [Q&A]
Hey there! There weren't any pre-made options that I could find for synchronizing state with the Zustand store.

Similar solutions:
https://github.com/reactjs/react-router-redux
https://github.com/IBM/mobx-react-router

Could you please explain how this may be added to Zustand?

**Comment by dai-shi** [comment]
I wonder if someone has experience with it.

Technically, one can set state to zustand store in route change handler.

**Reply by sandrig** [reply]
Thanks, but maybe someone has ready-made abstractions. Please don't close the thread

**Reply by dai-shi** [reply]
I didn't mean to close the thread. I meant to make the discussion active by calling out someone.

**Comment by sandrig** [comment]
You could probably try something like this

```js
// store.js
import create from 'zustand';
import { createBrowserHistory } from 'history';

const history = createBrowserHistory();

const routeMiddleware = (config) => (set, get, api) => {
  return {
    ...config,
    updateRoute: (route) => {
      set({ route });
      history.push(route);
    },
  };
};

const useStore = create(routeMiddleware((set) => ({
  // Your states are here
})));

// Global listener for updating storage state to route changes
history.listen((location) => {
  useStore.getState().updateRoute(location.pathname);
});

export { useStore, history };

```
